kzg/src/msm/bos_coster.rs

```diff
@@ -0,0 +1,217 @@
+use core::ops::Shl;
+use std::cmp::Ordering;
+use std::collections::BinaryHeap;
+use std::ops::Sub;
+
+use crate::{Fr, G1Affine, G1Fp, G1Mul, G1ProjAddAffine, Scalar256, G1};
+
+pub struct VariableBaseMSM;
+
+impl VariableBaseMSM {
+    // https://github.com/btclib-org/btclib was originally used as reference for this implementation
+    pub fn multi_scalar_mul<
+        TFr: Fr,
+        TG1: G1 + G1Mul<TFr>,
+        TG1Fp: G1Fp,
+        TG1Affine: G1Affine<TG1, TG1Fp>,
+        TProjAddAffine: G1ProjAddAffine<TG1, TG1Fp, TG1Affine>,
+    >(
+        points: &[TG1],
+        scalars: &[Scalar256],
+    ) -> TG1 {
+        debug_assert!(points.len() == scalars.len());
+
+        let mut work_points = Vec::with_capacity(points.len());
+        let mut heap: BinaryHeap<Pair> = BinaryHeap::with_capacity(points.len());
+
+        for (s, p) in scalars.iter().zip(points.iter()) {
+            if !s.is_zero() && !p.is_inf() {
+                let idx = work_points.len();
+                work_points.push(p.clone());
+                heap.push(Pair {
+                    scalar: *s,
+                    point_index: idx,
+                });
+            }
+        }
+
+        while heap.len() > 1 {
+            let Some(pair1) = heap.pop() else {
+                break;
+            };
+            let pair2_scalar = {
```

Comment:
I believe you can remove this block entirely, and just put all its content into loop body